File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:

        output = 0
        seen = {}


        def helper(idx):

            if idx in seen:
                return seen[idx]
            
            if idx >= len(s):
                return 1

            if s[idx] == "0":
                return 0
            
            
            
            ans = helper(idx + 1)
            if idx + 1 < len(s) and (s[idx] == "1" or (s[idx] == "2" and s[idx + 1] in "0123456")):
                ans += helper(idx + 2)
            
            seen[idx] = ans
            return seen[idx]
        
        return helper(0)

        

File: 0091-decode-ways/0091-decode-ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        if (s[0] == '0') return 0;
        
        int n = s.size();
        vector<int> dp(n + 1, 0);
        dp[0] = 1;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            int dig = s[i - 1] - '0';
            if (dig != 0) {
                dp[i] = dp[i - 1];
            }

            int prev_dig = s[i - 2] - '0';
            cout << prev_dig << endl;
            if (prev_dig == 1 or (prev_dig == 2 and dig <= 6)) {
                dp[i] += dp[i - 2];
            }
        }
        return dp[n];
    }
};











